README
MARVIN is a library for viewing and maintaining data in a PostgreSQL
database. It requires

          - GNU make 
          - postgresql
          - QT2

XPracMan is an application based on MARVIN. It is a Medical Practice
Management tool.

See INSTALL for information about installing PRAC



ChangeLog
2006-11-07  Source Code Management  <source@toshi2.mcmaster.wattle.id.au>

	* qtmarvin/qmvqueryconditionwidget.cpp (QmvQueryConditionWidget):
	Feature:: Disable auto-completion on search pattern

	* qtmarvin/qmvtable.cpp (setContentFromEditor): Feature::
	implement edit-prompting (set in mtat ... input method = "RWP")

	* qtmarvin/qmvsqlquery.cpp (noticeChanges): No longer active
	set-updates locally - just announce that the query has
	changed. This allows the qmvquerywidget object (at user-level) to
	control loading/cancelling of the set.
	(countEstimate): Feature:: return the count of rows associated
	with the currently constructed query.

	* qtmarvin/qmvrelationwidget.cpp (getCompositeQuery): New method.
	(formSet): Widen the possible form set which applies to the
	current set - extend from exact match to "starts-with"

	* qtmarvin/qmvquerywidget.cpp (initSimple): Feature:: use a
	query-controller to allow user to cancel queries.
	(getBaseQuery): new method
	(updateRelationSet): Feature:: new method to localise query for
	better user-control. Warn user if estimated rows in query exceed
	mtcl limit.

	* qtmarvin/qmvquerycondition.cpp (sqlCondition): Reduce the
	complexity of joined conditions and allow them to be joined with
	OR ("|")

	* qtmarvin/qmvdb.cpp (selectChild): apply user-filters to child list.

2006-10-12 22:24  source

	* qtmarvin/qmvtoolwindow.cpp, qtmarvin/qmvtoolwindow.h:
	  Enhancement: window to mirror main window, size and position

2006-10-04 22:56  source

	* qtmarvin/Makefile: add new class - qmvformeditfields

2006-10-02 11:52  source

	* qtmarvin/qmvmultilineedit.cpp: * qtmarvin/qmvmultilineedit.cpp
	  (QmvMultiLineEdit): add strong
	  focus policy

2006-10-02 11:52  source

	* qtmarvin/qmvform.cpp, qtmarvin/qmvform.h,
	  qtmarvin/qmvformeditfields.cpp, qtmarvin/qmvformeditfields.h:
	  Feature: add extensible method of editing specified fields on
	  forms.

2006-10-02 11:50  source

	* qtmarvin/qmvapplication.cpp, qtmarvin/qmvapplication.h: remove
	  references to documentation

2006-09-27 09:05  source

	* qtmarvin/qmvcanvasview.cpp: * qtmarvin/qmvcanvasview.cpp
	  (contentsMouseReleaseEvent): Notice
	  clicks which are not on objects - cab be used for controls.

2006-09-18 22:43  source

	* schema/marvin/create_metadata.sql: import uncommited changes
	  from
	  /usr/local/src/xpracman/src/schema/marvin/create_metadata.sql

2006-09-05 02:56  source

	* qtmarvin/qmvattribute.cpp, qtmarvin/qmvattribute.h,
	  qtmarvin/qmvtable.cpp, qtmarvin/qmvtable.h: Feature:
	  configurable edit-promting for field edits

2006-09-05 02:55  source

	* qtmarvin/.cdtproject, qtmarvin/.project: eclipse

2006-09-05 02:54  source

	* qtkugar/.cdtproject, qtkugar/.project: eclipse

2006-09-05 02:54  source

	* pgmarvin/.cdtproject, pgmarvin/.project: eclipse

2006-09-03 08:30  source

	* Makefile: lock in qt reference

2006-09-03 08:24  source

	* Makefile: Standardise default envars to SuSE

2006-09-02 02:38  source

	* qtmarvin/Makefile: Add GTAGS (used by doxygen)

2006-09-02 02:38  source

	* qtmarvin/Doxyfile: Add GTAGS (used by doxygen)

2006-09-02 02:28  source

	* qtkugar: ignore libqtkugar.a

2006-09-02 02:15  source

	* pgmarvin: ignore libpgmarvin.a

2006-09-02 02:01  source

	* pgmarvin/Makefile: Add GTAGS (used by doxygen)

2006-07-18 02:08  source

	* tools/tree_perms.sh: check and set permissions

2006-07-17 10:35  source

	* branches/XPRACMAN_4_2/medical/pracman/schema/prac/patn-triggers.sql,
	  schema/marvin/fn_column_default.sql: System Feature: replace
	  patn_feet_code with default value if set to null (eg during new
	  patient insert from appointment book)

2006-07-16 13:38  source

	* tools/vncusers-sbin: fix path to X11 socket

2006-07-11 10:42  source

	* qtmarvin/qmvtoolwindow.cpp, qtmarvin/qmvtoolwindow.h,
	  trunk/medical/pracman/xpracman/userpreferences,
	  trunk/medical/pracman/xpracman/xpracapptool.cpp: Feature:
	  configuration canvas zooming

2006-07-11 07:25  source

	* README: testing loginfo

2006-07-11 07:24  source

	* README: testing loginfo

2006-07-11 07:22  source

	* README: testing loginfo

2006-07-11 07:19  source

	* README: testing loginfo

2006-07-11 07:09  source

	* README: testing loginfo

2006-07-11 07:02  source

	* ChangeLog: testing loginfo - off

2006-07-11 06:00  source

	* Makefile: add new targets: cvstag ChangeLog
	  extend cvsupdate

2006-07-11 05:33  source

	* ChangeLog: regular update

2006-07-11 00:42  source

	* tools/Makefile, tools/vncusers-rc.d, tools/vncusers-sbin: -
	  improve cleaning process for stop and restart
	  - prevent overwriting sysconfig file

2006-07-10 07:59  source

	* tools/pg_admin.sh: make WEEKLIST an envar

2006-07-10 07:42  source

	* tools/xdump: modernise tar args

2006-07-10 07:40  source

	* tools/xdump: put excludes in external file

2006-07-10 06:56  source

	* tools/xdump: add md5 tags - incorrectly added to branch

2006-07-05 06:33  source

	* AUTHORS, ChangeLog, EXTENSION, INSTALL, Makefile:
	  MERGED_FROM_XPRACMAN_4_2

2006-07-05 04:56  source

	* tools/Makefile, tools/backups.sh, tools/dumpdata,
	  tools/dumpforms, tools/dumpforms.sh, tools/dumpmetadata,
	  tools/dumpschema, tools/mapshare.bat, tools/mirror,
	  tools/pg_admin.sh, tools/ps2dymo, tools/smb.conf,
	  tools/transfer_to_syncdir_csv.template, tools/txt2dymo,
	  tools/vncusers-rc.d, tools/vncusers-sbin,
	  tools/vncusers-sysconfig, tools/vncusers.vm, tools/xdump: Merge
	  from XPRACMAN_4_2

2006-07-02 08:31  source

	* qtmarvin/qmvroweditor.cpp, qtmarvin/qmvselect.cpp: *
	  qmvselect.cpp (init): Fix keyboard focus problem
	  
	  * qmvroweditor.cpp (QmvRowEditor): Fix keyboard focus problem

2006-06-15 03:41  source

	* qtmarvin/qmvrelationwidget.cpp, qtmarvin/qmvrelationwidget.h:
	  importSet(),addToSet(): allow quiet (no broadcast) additions to
	  set.

2006-06-14 02:18  source

	* qtmarvin/qmvrelationwidget.cpp: displayCell(): initialise et -
	  was causing sporadic display of colourdialog

2006-06-11 12:35  source


pgmarvin/misc.c
/*
 *     $Id: misc.c 917 2001-10-14 21:59:23Z source $
 * 
 *   Copyright 2000 X=X Computer Software Trust
 *    		  Kangaroo Ground Australia 3097
 * 
 * 
 *  This is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License published by
 *  the Free Software Foundation; either version 2, or (at your option)
 *  any later version.
 *  
 *  This software is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  
 *  You should have received a copy of the GNU General Public License
 *  along with this software; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *  
 *  Report problems and direct all questions to:
 *  
 *      Rex McMaster, rmcm@compsoft.com.au
 *  
 * 
 *  
 *  Revision Log
 *   
 *   $Log$
 *   Revision 1.3  2001/10/14 21:59:23  source
 *   add versioning variables
 *
 *   Revision 1.2  2000/11/05 07:51:20  source
 *   message_control_id(), timestamp(): transfer to hl7tool::hl7:
 *
 *   Revision 1.1  2000/10/23 03:38:19  source
 *   Version transferred from hl7tool
 *
 *   Revision 1.5  2000/10/04 22:59:57  source
 *   Remove cxref header
 *
 *   Revision 1.4  2000/09/28 02:21:23  source
 *   set_std_err(): new function
 *
 *   Revision 1.3  2000/09/28 02:04:49  source
 *   standardize format
 *
 *   Revision 1.2  2000/09/19 11:07:51  source
 *   syslog option
 *
 *   Revision 1.1  2000/09/17 06:34:31  source
 *   first cut
 *

 */


/*
 * ----------------------------------------------------------------------
 * DESCRIPTION: Miscellaneous functions
 *
 * ----------------------------------------------------------------------
 */



/*
 * ----------------------------------------------------------------------
 * INCLUDES:
 * ----------------------------------------------------------------------
 */

#include <syslog.h>
#include <stdarg.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <time.h>
#include "misc.h"

/*
 * ----------------------------------------------------------------------
 * DEFINES:
 * ----------------------------------------------------------------------
 */

/*
 * ----------------------------------------------------------------------
 * LOCAL VARIABLES:
 * ----------------------------------------------------------------------
 */
    
static char misc_version[] = "$Id: misc.c 917 2001-10-14 21:59:23Z source $";
static enum std_error local_std_error; /*+ Error destination +*/
/*
 * ----------------------------------------------------------------------
 * LOCAL FUNCTION PROTOTYPES:
 * ----------------------------------------------------------------------
 */

/*
 * ----------------------------------------------------------------------
 * LOCAL FUNCTIONS:
 * ----------------------------------------------------------------------
 */

/*
 * ----------------------------------------------------------------------
 * GLOBAL FUNCTIONS:
 * ----------------------------------------------------------------------
 */

/*
 * ----------------------------------------------------------------------
 * FUNCTION:    set_std_err
 * DESCRIPTION: Set the static global variable std_error
 * ARGS:        i
 * RETURNS:     
 * ----------------------------------------------------------------------
 */

void
set_std_err(i)
enum std_error i;
{
    local_std_error = i;
    
} /* set_std_err */

/*
 * ----------------------------------------------------------------------
 * FUNCTION:    misc_warnx
 * DESCRIPTION: 
 * ARGS:        format ...
 * RETURNS:     
 * ----------------------------------------------------------------------
 */

void
misc_warnx(char *fmt, ...)      /*+ Wrapper for warnx +*/
{
    va_list ap;

    va_start(ap, fmt);

    if (local_std_error == USE_SYSLOG) {
        vsyslog(LOG_INFO, fmt, ap);
    } else {
        vfprintf(stderr, fmt, ap);
        fprintf(stderr, "\n");
        fflush(stderr);
    }
    va_end(ap);

} /* misc_warnx */

/*
 * ----------------------------------------------------------------------
 * FUNCTION:    misc_errx
 * DESCRIPTION: 
 * ARGS:        format ...
 * RETURNS:     
 * ----------------------------------------------------------------------
 */

void
misc_errx(int eval, char *fmt, ...)
{
    va_list ap;

    va_start(ap, fmt);

    if (local_std_error == USE_SYSLOG) {
        vsyslog(LOG_INFO, fmt, ap);
    } else {
        vfprintf(stderr, fmt, ap);
        fprintf(stderr, "\n");
        fflush(stderr);
    }
    va_end(ap);

    exit(eval);

} /* misc_errx */

